framework/src/media/HttpInputDataSource.cpp

```diff
@@ -0,0 +1,279 @@
+	}
+
+	if (mHttpStream == nullptr) {
+		mHttpStream = HttpStream::create();
+		if (mHttpStream == nullptr) {
+			meddbg("mHttpStream is nullptr!\n");
+			return false;
+		}
+	}
+
+	std::unique_lock<std::mutex> lock(mMutex);
+	mIsHeaderReceived = false;
+	mIsDataReceived = false;
+
+	mThread = std::thread([this]() {
+		medvdbg("download thread enter!\n");
+		//mHttpStream->addHeader("Icy-MetaData:1"); // not support now
+		mHttpStream->setHeaderCallback(HeaderCallback, (void *)this);
+		mHttpStream->setWriteCallback(WriteCallback, (void *)this);
+		if (!mHttpStream->download(mUrl)) {
+			medwdbg("download failed or terminated!\n");
+			// TODO: send network error code to upper layer later
+		}
+
+		mBufferWriter->setEndOfStream();
+		medvdbg("download thread exit!\n");
+	});
+
+	// wait for Content-Type header
+	if (!mCondv.wait_for(lock, WAIT_HEADER_TIMEOUT, [=]{ return mIsHeaderReceived; })) {
+		medwdbg("download:: wait header timeout!\n");
+		mBufferWriter->setEndOfStream();
+		return false;
+	}
+
+	audio_type_t audioType = AUDIO_TYPE_UNKNOWN;
+	if ((mContentType.find(AAC_MIME_TYPE) != std::string::npos) ||
+		(mContentType.find(AACP_MIME_TYPE) != std::string::npos)) {
+		audioType = AUDIO_TYPE_AAC;
+	} else if (mContentType.find(MPEG_MIME_TYPE) != std::string::npos) {
+		audioType = AUDIO_TYPE_MP3;
+	} else if (mContentType.find(OPUS_MIME_TYPE) != std::string::npos) {
+		audioType = AUDIO_TYPE_OPUS;
+	} else {
+		meddbg("Unsupported internet media type: %s\n", mContentType.c_str());
+	}
+
+	switch (audioType) {
+	case AUDIO_TYPE_MP3:
+	case AUDIO_TYPE_AAC: {
+		// wait for audio stream data
+		if (!mCondv.wait_for(lock, WAIT_DATA_TIMEOUT, [=]{ return mIsDataReceived; })) {
+			medwdbg("download:: wait audio data timeout!\n");
+			mBufferWriter->setEndOfStream();
+			return false;
+		}
+
+		size_t templen = mBufferReader->sizeOfData();
+		assert(templen != 0);
+		unsigned char *tempbuf = new unsigned char[templen];
+		assert(tempbuf != nullptr);
+		size_t dlen = mBufferReader->dump(0, tempbuf, templen);
+		assert(templen == dlen);
+
+		unsigned int channel;
+		unsigned int sampleRate;
+		bool ret = utils::header_parsing(tempbuf, dlen, audioType, &channel, &sampleRate, NULL);
+		delete[] tempbuf;
+
+		if (!ret) {
+			meddbg("header parsing failed\n");
+			mBufferWriter->setEndOfStream();
+			return false;
+		}
+
+		medvdbg("audioType:%d,setSampleRate:%u,setChannels:%u\n", (int)audioType, sampleRate, channel);
+		setAudioType(audioType);
+		setSampleRate(sampleRate);
+		setChannels(channel);
+		break;
+	}
+
+	default:
+		/* unsupported audio type */
+		meddbg("HttpInputDataSource::open, unsupported audio type %d\n", (int)audioType);
+		mBufferWriter->setEndOfStream();
+		return false;
+	}
+
+	medvdbg("HttpInputDataSource::open! exit\n");
+	return true;
+}
+
+bool HttpInputDataSource::close()
+{
+	medvdbg("HttpInputDataSource::close enter\n");
+	if (mBufferWriter) {
+		mBufferWriter->setEndOfStream();
+	}
+
+	if (mThread.joinable()) {
```

Comment:
In this situation, you will wait for the curl download to complete.
Is there a way to immediately terminate the thread?

Reply:
once mBufferWriter->setEndOfStream() is called, streamwriter would always return 0, 
then curl_perform will end and return, and the download thread would be terminated at once.